### PR TITLE
CLI [5397]

### PR DESCRIPTION
**Pull request by julianbermudez**
This pull request adds a CLI to the Agent executable.
The CLI is base on [CLI11](https://github.com/CLIUtils/CLI11).

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/170/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/173/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/175/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/178/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/178/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/181/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/180/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/183/)

File: include/uxr/agent/utils/CLI.hpp [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+                rv = server_->run();
+            }
+        }
+
+        return rv;
+    }
+
+private:
+    BaudrateOpt baudrate_opt_;
+    CommonOpts common_opts_;
```
Comment (BorjaOuterelo) [inline include/uxr/agent/utils/CLI.hpp]: Could this be in the base class, `Subcommand`?.

**Comment by BorjaOuterelo** [comment]
Running the application without commands should print help text.

File: include/uxr/agent/Agent.hpp [hunk 2]
```diff
@@ -367,6 +367,10 @@ class Agent
             size_t len,
             OpResult& op_result);
 
+    /**********************************************************************************************
+     * Config.
+     **********************************************************************************************/
+    UXR_AGENT_EXPORT static void set_verbose_level(uint8_t verbose_level);
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply include/uxr/agent/Agent.hpp]: Fixed!

File: include/uxr/agent/logger/Logger.hpp [hunk 3]
```diff
@@ -108,9 +108,9 @@
 #endif
 
 #ifdef PROFILE_LOGGER
-#define UXR_AGENT_LOG_CRITICAL(...) SPDLOG_CRITICAL(UXR_STATUS_FORMAT __VA_ARGS__)
+#define UXR_AGENT_LOG_CRITICAL(...) SPDLOG_CRITICAL(UXR_STATUS_FORMAT __VA_ARGS__); exit(1)
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply include/uxr/agent/logger/Logger.hpp]: Changed!

File: src/cpp/client/ProxyClient.cpp [hunk 4]
```diff
@@ -40,6 +40,13 @@ ProxyClient::ProxyClient(
 {
     switch (middleware_kind)
     {
+        case Middleware::Kind::NONE:
+        {
+            UXR_AGENT_LOG_CRITICAL(
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/client/ProxyClient.cpp]: Changed!

File: src/cpp/transport/udp/UDPServerLinux.cpp [hunk 5]
```diff
@@ -42,6 +42,11 @@ UDPServer::UDPServer(
 #endif
 {}
 
+UDPServer::~UDPServer()
+{
+    stop();
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply src/cpp/transport/udp/UDPServerLinux.cpp]: I moved the `stop()` call to the base class `Server` and catch the exception.

File: include/uxr/agent/utils/CLI.hpp (hunk 1 above)
Comment (julianbermudez) [reply include/uxr/agent/utils/CLI.hpp]: It requires a refactor, I add a TODO.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/191/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/194/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/194/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/197/)

File: include/uxr/agent/logger/Logger.hpp [hunk 6]
```diff
@@ -109,7 +109,7 @@
 #endif
 
 #ifdef PROFILE_LOGGER
-#define UXR_AGENT_LOG_CRITICAL(X, Y, ...) SPDLOG_CRITICAL(UXR_STATUS_FORMAT Y, X, __VA_ARGS__); exit(1)
+#define UXR_AGENT_LOG_CRITICAL(X, Y, ...) SPDLOG_CRITICAL(UXR_STATUS_FORMAT Y, X, __VA_ARGS__); std::exit(EXIT_FAILURE)
 #else
 #define UXR_AGENT_LOG_CRITICAL(...) exit(1)
```
[the remark this answers is not in the document]
Comment (julianbermudez) [reply include/uxr/agent/logger/Logger.hpp]: Fixed!

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Linux/195/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro%20XRCE-DDS%20Agent%20Manual%20Windows/198/)

**Pull request merged**